### PR TITLE
LF: shortcup failure in TypeOrdering

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR, we shortcut failure, before comparison of structural record has always a linear complexity (even in case of failure of the first component).

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala [hunk 1]
```diff
@@ -38,52 +29,72 @@ object TypeOrdering extends Ordering[Ast.Type] {
...
+    @inline
+    def step(tuple: (Ast.Type, Ast.Type)): Unit =
+      tuple match {
+        case (Ast.TBuiltin(x), Ast.TBuiltin(y)) =>
+          diff = builtinTypeRank(x) compare builtinTypeRank(y)
+        case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
+          diff = con1 compare con2
+        case (Ast.TNat(x), Ast.TNat(y)) =>
+          diff = x compare y
+        case (Ast.TStruct(ImmArray((xField, xType))), Ast.TStruct(ImmArray((yField, yType)))) =>
```
Comment (cocreature) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala:74]: Maybe that’s just me but splitting multi-field structs into single-field structs seems quite convoluted and at least deserves a comment. This is also the only case where you actually make use of the fact that the elements of your stacks are iterators. Not quite sure how to clean it up, is there a reason why we wrote the spec to first compare individual field types before comparing the number of fields? The latter seems more intuitive and would also make the implementation simpler. I guess it’s still in 1.dev so if we want to change it now would be the time.

**Review by cocreature — APPROVED** [review]
LGTM with one comment, thank you!

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -2005,17 +2005,17 @@ types that satisfies the following rules::
     n <ₜ ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩
 
     fᵢ comes lexicographically before gᵢ
-  ——————————————————————————————————————————————————— TypeOrderStructFieldName
-    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
-      ⟨ f₁ : σ₁, …, fᵢ₋1 : σᵢ₋₁, gᵢ : σᵢ, …, gₙ : σₙ ⟩
+  ——————————————————————————————————————————————————— TypeOrderStructFieldNameDiff
+    ⟨  f₁ : τ₁, …, fᵢ₋₁ : τᵢ₋₁, fᵢ : τᵢ, …, fₘ : τₘ ⟩ <ₜ
+        ⟨  f₁ : σ₁, …, fᵢ₋₁ : σᵢ₋₁, gᵢ : σᵢ, …, gₘ : σₘ⟩
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:2010]: Wouldn't it be easier to say something like
```
  [f_1, ..., f_m] is lexicographically less than [g_1, ..., g_n]
------------------------------------------------------------------
  <f_1: s_1, ..., f_m: s_m> <_t <g_1: t_1, ..., g_n: t_n>
```
?

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -2005,17 +2005,17 @@ types that satisfies the following rules::
...
 
-  ——————————————————————————————————————————————————— TypeOrderStructFieldNumber
-    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
-      ⟨ f₁ : τ₁, …, fₘ : τₘ, fₘ₊₁ : τₘ₊₁ ⟩
+  ——————————————————————————————————————————————————— TypeOrderStructFieldNameShorter
+    ⟨  f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
+         ⟨ f₁ : σ₁, …, fₘ : σₘ, fₘ₊₁ : σₘ₊₁, …, fₘ : σₘ⟩⟩
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:2014]: The second struct contains the field `f_m` twice. I assume that second one should be `f_n`? However, this rule would be unnecessary if we follow my suggestion above.

File: daml-lf/spec/daml-lf-1.rst (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:2010]: It is simpler. We agree with @cocreature to go this way. 
Note the resulting order is not the same. It is somehow less natural as it does not behave like map.

**Comment by remyhaemmerle-da** [comment]
waits #7241, #7314, and #7319.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala]: changed.

File: daml-lf/spec/daml-lf-1.rst (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: fixed in other PR.

**Comment by remyhaemmerle-da** [comment]
The PR was reworked quite extensively.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala [hunk 4]
```diff
@@ -1,22 +1,71 @@
...
+    }
+
+    @inline
+    def pop(): Unit = {
+      stackX = stackX.tail
+      stackY = stackY.tail
+    }
+
+    @inline
+    def compareNames(xs: Iterator[Ref.Name], ys: Iterator[Ref.Name]): Unit = {
```
Comment (hurryabit) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala:33]:
```suggestion
    def compareNamesLexicographically(xs: Iterator[Ref.Name], ys: Iterator[Ref.Name]): Unit = {
```

**Review by hurryabit — APPROVED** [review]
Very nice. Thank you.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala [hunk 5]
```diff
@@ -1,22 +1,71 @@
...
+    def compareNamesLexicographically(xs: Iterator[Ref.Name], ys: Iterator[Ref.Name]): Unit = {
+      while (diff == 0 && xs.hasNext && ys.hasNext) diff = xs.next() compare ys.next()
+      if (diff == 0) diff = xs.hasNext compare ys.hasNext
+    }
+
+    @inline
+    def step(tuple: (Type, Type)): Unit =
+      tuple match {
+        case (Ast.TBuiltin(b1), Ast.TBuiltin(b2)) =>
+          diff = builtinTypeIdx(b1) compare builtinTypeIdx(b2)
```
Comment (remyhaemmerle-da) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala:42]:
```suggestion
          diff = builtinTypeIdx(b1) compareTo builtinTypeIdx(b2)
```

**Pull request merged**